File: day4/swift/main.swift
// AoC day 4

import Foundation

func readFile(filename: String) -> [String] {
    do {
        let text = try String(contentsOfFile: filename, encoding: .ascii)
        let lines : [String] = text.components(separatedBy: "\n")
        return lines
    } catch let _ as NSError{
        return[]
    }
}

func isValidHex(input: String) -> Bool {
    for c in input {
        if !c.isHexDigit {
            return false
        }
    }
    return true
}

class Passport {
    var birthYear = ""
    var issueYear = ""
    var expireYear = ""
    var height = ""
    var hairColor = ""
    var eyeColor = ""
    var passportId = ""
    var countryId = ""
    
    func naiveValidate() -> Bool {
        var fieldsPresent = 0;
        if !birthYear.isEmpty {
            fieldsPresent += 1
        }
        if !issueYear.isEmpty {
            fieldsPresent += 1
        }
        if !expireYear.isEmpty {
            fieldsPresent += 1
        }
        if !height.isEmpty {
            fieldsPresent += 1
        }
        if !hairColor.isEmpty {
            fieldsPresent += 1
        }
        if !eyeColor.isEmpty {
            fieldsPresent += 1
        }
        if !passportId.isEmpty {
            fieldsPresent += 1
        }
        if !countryId.isEmpty {
            fieldsPresent += 1
        }
        
        let allFieldsPresent = fieldsPresent == 8
        let missingCid = fieldsPresent == 7 && countryId.isEmpty
        return allFieldsPresent || missingCid
    }
    
    func validRange(input: String, start: Int, end: Int) -> Bool {
        if let y = Int(input) {
            return y >= start && y <= end
        }
        return false
    }
    func validate() -> Bool {
        let birthYearValid = validRange(input: birthYear, start: 1920, end: 2002)
        let issueYearValid = validRange(input: issueYear, start: 2010, end: 2020)
        let expireYearValid = validRange(input: expireYear, start: 2020, end: 2030)
        
        var heightValid = false
        let isCm = height.contains("cm")
        let isIn = height.contains("in")
        if isCm {
            let tmp = height.replacingOccurrences(of: "cm", with: "")
            if let x = Int(tmp) {
                heightValid = x >= 150 && x <= 193
            }
        }
        if isIn {
            let tmp = height.replacingOccurrences(of: "in", with: "")
            if let x = Int(tmp) {
                heightValid = x >= 59 && x <= 76
            }
        }
        var hairColorValid = false
        if (hairColor.starts(with: "#")) {
            hairColorValid = isValidHex(input: hairColor.replacingOccurrences(of: "#", with: ""))
        }
        
        let validColors = ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"]
        let eyeColorValid = validColors.contains(eyeColor)
        
        var passportIdValid = false
        if (passportId.count == 9) {
            if let _ = Int(passportId) {
                passportIdValid = true
            }
        }
        
        return birthYearValid &&
        issueYearValid &&
        expireYearValid &&
        heightValid &&
        hairColorValid &&
        eyeColorValid &&
        passportIdValid
    }
}

func parseData(input: [String]) -> Passport {
    let res = Passport()
    for item in input {
        let pairs = item.split(separator: " ")
        for pair in pairs {
            let kvp = pair.split(separator: ":")
            let key = String(kvp[0]).lowercased()
            let value = String(kvp[1])
            switch key {
            case "byr":
                res.birthYear = value
            case "iyr":
                res.issueYear = value
            case "eyr":
                res.expireYear = value
            case "hgt":
                res.height = value
            case "ecl":
                res.eyeColor = value
            case "hcl":
                res.hairColor = value
            case "pid":
                res.passportId = value
            case "cid":
                res.countryId = value
            default:
                break
            }
        }
    }
    return res
}

func parsePassports(lines: [String]) -> [Passport] {
    var passports: [Passport] = []
    var input: [String] = []
    for line in lines {
        if line.isEmpty {
            passports.append(parseData(input: input));
            input = []
        }
        else {
            input.append(line)
        }
    }
    if !input.isEmpty {
        passports.append(parseData(input: input));
    }
    return passports
}


let lines = readFile(filename: "/Users/mikaelstalvik/Downloads/aoc4f.txt")

let passports = parsePassports(lines: lines)

var valid = 0
for passport in passports {
    if passport.naiveValidate() {
        valid += 1
    }
}
print("Valid: \(valid) shall be 192")

valid = 0
for passport in passports {
    if passport.validate() {
        valid += 1
    }
}
print("Valid: \(valid) shall be 101")

File: day4/cpp/main.cpp
#include <algorithm>
#include <iostream>
#include <string>
#include <utility>
#include <vector>
#include <fstream>
#include <sstream>
#include <tuple>

using namespace std;

static tuple<bool, int> to_int(const string& s)
{
    if (s.empty()) return make_tuple(false, -1);
	try
	{
        auto t = stoi(s);
        return make_tuple(true, t);
	}
	catch(...)
	{
        return make_tuple(false, -1);
	}
}

std::vector<std::string> split(const std::string& text, const char delim) {
    std::string line;
    std::vector<std::string> vec;
    std::stringstream ss(text);
    while (std::getline(ss, line, delim)) {
        vec.push_back(line);
    }
    return vec;
}

string replace(string str, const string& str2, const string& str3)
{
    return str.replace(str.find(str2), str2.length(), str3);
}

bool is_in_range(const string& s, const int v1, const int v2)
{
    bool valid;
    int year;
    tie(valid, year) = to_int(s);
    return valid && year >= v1 && year <= v2;
}

class Passport
{
public:
    string birth_year;
    string issue_year;
    string expire_year;
    string height;
    string hair_color;
    string eye_color;
    string passport_id;
    string country_id;

	bool naive_is_valid() const
	{
        auto fields_present = 0;
        if (!empty(birth_year)) fields_present++;
        if (!empty(issue_year)) fields_present++;
        if (!empty(expire_year)) fields_present++;
        if (!empty(height)) fields_present++;
        if (!empty(hair_color)) fields_present++;
        if (!empty(eye_color)) fields_present++;
        if (!empty(passport_id)) fields_present++;
        if (!empty(country_id)) fields_present++;
        const auto all_fields_present = fields_present == 8;
        const auto missing_cid = fields_present == 7 && empty(country_id);
        return all_fields_present || missing_cid;
    }
	bool is_valid() const
	{
        return birth_year_valid() && issue_year_valid() && expire_year_valid() && height_valid() && hair_color_valid() && eye_color_valid() && passport_id_valid();
	}
	
private:
	bool birth_year_valid() const
	{
        return is_in_range(birth_year, 1920, 2002);
	}
    bool issue_year_valid() const
    {
        return is_in_range(issue_year, 2010, 2020);
    }
    bool expire_year_valid() const
    {
        return is_in_range(expire_year, 2020, 2030);
    }
	bool height_valid() const
	{
		const auto is_cm = static_cast<int>(height.find("cm", 0)) >= 0;
        const auto is_in = static_cast<int>(height.find("in", 0)) >= 0;
		if (is_cm)
		{
			const auto tmp = replace(height, "cm", "");
            return is_in_range(tmp, 150, 193);
		}
		if (is_in)
		{
            const auto tmp = replace(height, "in", "");
            return is_in_range(tmp, 59, 76);
        }
        return false;
	}
	bool hair_color_valid() const
	{
		if (hair_color.rfind('#', 0) == 0)
		{
            auto tmp = replace(hair_color, "#", "");
            if (std::all_of(tmp.begin(), tmp.end(), ::isxdigit)) return true;
            return false;
		}
        return false;
    }
	bool eye_color_valid() const
	{
        auto valid_colors = vector<string>{ "amb", "blu", "brn", "gry", "grn", "hzl", "oth" };
        auto key = eye_color;
        std::transform(key.begin(), key.end(), key.begin(),
            [](unsigned char c) { return std::tolower(c); });
        return count(valid_colors.begin(), valid_colors.end(), key);
	}
	bool passport_id_valid() const
	{
		if (passport_id.length() == 9)
		{
            bool valid;
            int v;
            tie(valid, v) = to_int(passport_id);
            return valid;
        }
        return false;
	}
};

static vector<string> read_file(const string& filename)
{
    ifstream inf(filename);
    vector<string> lines;
    string line;
    for (unsigned int i = 1; getline(inf, line); ++i)
        lines.emplace_back(line);
    return lines;
}

static Passport parse_data(const vector<string>& input)
{
    auto res = Passport();
	for(const auto& item : input)
	{
        auto pairs = split(item, ' ');
		for(const auto& pair : pairs)
		{
            auto kvp = split(pair, ':');
            auto key = kvp[0];
            std::transform(key.begin(), key.end(), key.begin(),
                [](unsigned char c) { return std::tolower(c); });
            const auto value = kvp[1];
            if (key == "byr") res.birth_year = value;
            if (key == "iyr") res.issue_year = value;
            if (key == "eyr") res.expire_year = value;
            if (key == "hgt") res.height = value;
            if (key == "ecl") res.eye_color = value;
            if (key == "hcl") res.hair_color = value;
            if (key == "pid") res.passport_id = value;
            if (key == "cid") res.country_id = value;
        }
	}
    return res;
}

static vector<Passport> passports;

static void parse_passports(const vector<string>& lines)
{
    passports.clear();
	
    auto input = vector<string>();
	for(const auto& line : lines)
	{
		if (empty(line))
		{
            passports.emplace_back(parse_data(input));
            input.clear();
		} else
		{
            input.emplace_back(line);
		}
	}
	if (!input.empty()) passports.emplace_back(parse_data(input));
}

int main()
{
	const auto lines = read_file("c:\\temp\\aoc4f.txt");

    parse_passports(lines);    

    auto valid = 0;
	for(const auto& passport : passports)
	{
        if (passport.naive_is_valid()) valid++;
	}
    cout << "Valid: " << valid << " shall be 192\n";
    
    valid = 0;
	for(const auto& passport : passports)
	{
        if (passport.is_valid()) valid++;
	}
    cout << "Valid: " << valid << " shall be 101\n";
    passports.clear();
}
